subiquity/server/controllers/zdev.py

```diff
@@ -607,20 +643,60 @@ def __init__(self, app):
                 zdevinfos = [ZdevInfo.from_row(row) for row in devices]
             self.zdevinfos = OrderedDict([(i.id, i) for i in zdevinfos])
 
+    def load_autoinstall_data(self, data: ZdevAi) -> None:
+        self.ai_data = data
+
+    @with_context()
+    async def apply_autoinstall_config(self, context) -> None:
+        if self.zdev_handling_task is not None:
+            await self.zdev_handling_task
+
+    def start(self) -> None:
+        if self.ai_data:
+            self.zdev_handling_task = schedule_task(self.handle_zdevs())
+
+    def make_autoinstall(self) -> ZdevAi:
+        # Small "optimization" to avoid producing a config that enables or
+        # disables a given device multiple times in a row.
+        return [x for x, _ in itertools.groupby(self.done_actions)]
+
+    async def handle_zdevs(self) -> None:
+        for item in self.ai_data:
+            if item["state"] == "enabled":
+                await self.chzdev("enable", self.zdevinfos[item["id"]])
+            elif item["state"] == "disabled":
+                await self.chzdev("disable", self.zdevinfos[item["id"]])
+            else:
+                raise ValueError('state must be "enabled" or "disabled"')
+
     def interactive(self):
         if self.app.base_model.filesystem.bootloader != Bootloader.NONE:
             return False
         return super().interactive()
 
-    async def chzdev_POST(self, action: str, zdev: ZdevInfo) -> List[ZdevInfo]:
+    async def chzdev(
+        self, action: Literal["enable", "disable"], zdev: ZdevInfo
+    ) -> None:
+        if action == "enable":
+            on = True
+            state = "enabled"
+        elif action == "disable":
+            on = False
+            state = "disabled"
+        else:
+            raise ValueError("action must be 'enable' or 'disable'")
```

Comment:
```suggestion
            raise AutoinstallError("action must be 'enable' or 'disable'")
```

Reply:
`chzdev` is also called by non-autoinstall code. Maybe we catch the ValueError and wrap it in an AutoinstallError in autoinstall code.